clang/include/clang/CIR/Dialect/IR/CIRAttrs.td

```diff
@@ -1249,7 +1250,150 @@ def GlobalAnnotationValuesAttr : CIR_Attr<"GlobalAnnotationValues",
   let genVerifyDecl = 1;
 }
 
+
+//===----------------------------------------------------------------------===//
+// TBAAAttr
+//===----------------------------------------------------------------------===//
+
 def CIR_TBAAAttr : CIR_Attr<"TBAA", "tbaa", []> {
+  let summary = "CIR dialect TBAA base attribute";
+}
+
+//===----------------------------------------------------------------------===//
+// TBAATypeDescriptorAttr
+//===----------------------------------------------------------------------===//
+
+def CIR_TBAAMemberAttr : CIR_Attr<"TBAAMember", "tbaa_member", []> {
+  let summary = "Attribute representing a member of a TBAA structured type.";
+
+  let parameters = (ins "TBAAAttr":$typeDesc,
+                        "int64_t":$offset,
+                        "int64_t":$size);
+
+  let builders = [
+    AttrBuilderWithInferredContext<(ins "TBAAAttr":$typeDesc,
+                                        "int64_t":$offset,
+                                        "int64_t":$size), [{
+      return $_get(typeDesc.getContext(), typeDesc, offset, size);
+    }]>
+  ];
+  let assemblyFormat = "`<` params `>`";
+}
+
+def CIR_TBAAMemberAttrArray : ArrayRefParameter<"TBAAMemberAttr"> {
+  let summary = "Array of TBAAMemberAttr attributes.";
+
+  let printer = [{
+    $_printer << '{';
+    llvm::interleaveComma($_self, $_printer, [&](TBAAMemberAttr attr) {
+        $_printer.printStrippedAttrOrType(attr);
+    });
+    $_printer << '}';
+  }];
+
+  let parser = [{
+    [&]() -> llvm::FailureOr<llvm::SmallVector<TBAAMemberAttr>> {
+        using Result = llvm::SmallVector<TBAAMemberAttr>;
+        if ($_parser.parseLBrace())
+            return mlir::failure();
+        llvm::FailureOr<Result> result = mlir::FieldParser<Result>::parse($_parser);
+        if (failed(result))
+            return mlir::failure();
+        if ($_parser.parseRBrace())
+            return mlir::failure();
+        return result;
+    }()
+  }];
+}
+
+class CIR_TBAATypeDescriptorAttr<string name, string attrMnemonic>: CIR_Attr<name, attrMnemonic, [], "TBAAAttr"> {
+  let summary = "Base class for TBAA type descriptor attributes.";
```

Comment:
Why do you need this base class and what's the intent to compose on top of it? Same for CIR_TBAAAttr, can you elaborate?

Reply:
Still waiting for an answer for this since it affects my review of https://github.com/llvm/clangir/pull/1220

Reply:
I apologize for overlooking this question.

I make a distinction between scalar types and struct types. The use of `class CIR_TBAATypeDescriptorAttr` serves to constrain both `CIR_TBAAScalarTypeDescriptorAttr` and `CIR_TBAAStructTypeDescriptorAttr` as TBAA type descriptor attributes. If `CIR_TBAATypeDescriptorAttr` is redundant, we can remove it.

The purpose of using `CIR_TBAAAttr` as a base class in C++ is to provide a unified way to handle both `CIR_TBAAScalarTypeDescriptorAttr` and `CIR_TBAAStructTypeDescriptorAttr`.

Reply:
I haven't seen them get used in the PR though, hence the question. If are actually using it then fine, otherwise best wait until it's used to introduce. Let's move any discussion over the actual patches that might get landed, thanks for the reply!